File: 2022/15/rust/src/main.rs
pub mod map;
pub mod sensor;

use map::Map;
use regex::Regex;

// --- Constants --
const INPUT_FILE: &'static str = "../input/input.txt";
const PARSER_REGEX: &'static str = "Sensor at x=(?:(-?\\d+)), y=(?:(-?\\d+)): \
    closest beacon is at x=(?:(-?\\d+)), y=(?:(-?\\d+))";
const PUZZLE1_LINE: isize = 2_000_000;
const PUZZLE2_MIN: isize = 0;
const PUZZLE2_MAX: isize = 4_000_000;

// --- Helper functions ---
fn in_limit(x: isize, y: isize) -> bool {
    x >= PUZZLE2_MIN && x <= PUZZLE2_MAX && y >= PUZZLE2_MIN && y <= PUZZLE2_MAX
}

fn calculate_freq(x: isize, y: isize) -> u64 {
    let x = x as u64;
    let y = y as u64;
    let m = PUZZLE2_MAX as u64;
    x * m + y
}

// --- Main function ---
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut map = Map::new();

    // --- Read and parse the input file ---
    let text = std::fs::read_to_string(INPUT_FILE)?;
    let regex = Regex::new(PARSER_REGEX)?;
    for cap in regex.captures_iter(&text) {
        let sx: isize = cap[1].parse()?;
        let sy: isize = cap[2].parse()?;
        let bx: isize = cap[3].parse()?;
        let by: isize = cap[4].parse()?;
        map.add(sx, sy, bx, by);
    }

    // --- Search Max and Min ---
    let mut min = isize::max_value();
    let mut max = isize::min_value();
    let mut range = 0;
    for s in map.sensors.values() {
        min = std::cmp::min(min, s.x);
        max = std::cmp::max(max, s.x);
        range = std::cmp::max(range, s.range);
    }
    min -= range as isize;
    max += range as isize;

    // --- Puzzle 1 --
    let mut total: u64 = 0;
    for x in min..=max {
        if map.is_in_sensor_range(x, PUZZLE1_LINE) && map.is_empty(x, PUZZLE1_LINE) {
            total += 1;
        }
    }
    println!("1. Positions that cannot contain a beacon: {}", total);

    // --- Puzzle 2 ---
    let mut freq = 0;
    for s in map.sensors.values() {
        let r = s.range as isize;
        for i in 0..=r {
            // North -> East
            let x = s.x + i;
            let y = s.y - r + i - 1;
            if in_limit(x, y) && !map.is_in_sensor_range(x, y) {
                freq = calculate_freq(x, y); break;
            }

            // East -> South
            let x = s.x + r - i + 1;
            let y = s.y + i;
            if in_limit(x, y) && !map.is_in_sensor_range(x, y) {
                freq = calculate_freq(x, y); break;
            }

            // South -> West
            let x = s.x - i;
            let y = s.y + r - i + 1;
            if in_limit(x, y) && !map.is_in_sensor_range(x, y) {
                freq = calculate_freq(x, y); break;
            }

            // West -> North
            let x = s.x - r + i - 1;
            let y = s.y - i;
            if in_limit(x, y) && !map.is_in_sensor_range(x, y) {
                freq = calculate_freq(x, y); break;
            }
        }
        if freq > 0 { break; }
    }
    println!("2. Frequency: {}", freq);

    Ok(())
}

File: 2022/15/python/main.py
import re
from lib import Map

# --- Constants --
INPUT_FILE = "../input/input.txt"
REGEX = r"Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)"
PUZZLE1_LINE = 2000000
PUZZLE2_MIN = 0
PUZZLE2_MAX = 4000000

# --- Helper functions ---
def in_limit(x, y):
    return PUZZLE2_MIN <= x <= PUZZLE2_MAX and PUZZLE2_MIN <= y <= PUZZLE2_MAX

def calculate_freq(x, y):
    return x * PUZZLE2_MAX + y

# --- Variables ---
_map = Map()
_min = 2 ** 31 - 1
_max = 0
_range = 0

# --- Read and parse the input file ---
file = open(INPUT_FILE, "r")
text = file.read().strip()
for cap in re.findall(REGEX, text):
    sx = int(cap[0]); sy = int(cap[1])
    bx = int(cap[2]); by = int(cap[3])
    _map.add(sx, sy, bx, by)
    _min = min(_min, sx)
    _max = max(_max, sx)
    _range = max(_range, abs(sx - bx) + abs(sy - by))

_min -= _range
_max += _range + 1

# --- Puzzle 1 ---
total = 0
for x in range(_min, _max):
    if x % 10000 == 0:
        prc = round((x - _min) / (_max - _min)  * 100)
        print(f"\rProgress: {prc}%  ", end="")
    if _map.is_in_sensor_range(x, PUZZLE1_LINE) and _map.is_empty(x, PUZZLE1_LINE):
        total += 1

print(f"\r1. Positions that cannot contain a beacon: {total:,d}")

# --- Puzzle 2 ---
freq = 0
total = 0
for p in _map.sensors:
    total += 1
    prc = round(total / len(_map.sensors) * 100)
    print(f"\rProgress: {prc}%  ", end="")

    s = _map.sensors[p]
    r = s.range
    for i in range(r + 1):
        # North -> East
        x = s.x + i
        y = s.y -r + i - 1
        if in_limit(x, y,) and _map.is_in_sensor_range(x, y) == False:
            freq = calculate_freq(x, y)
            break

        # East -> South
        x = s.x + r - i + 1
        y = s.y + i
        if in_limit(x, y) and _map.is_in_sensor_range(x, y) == False:
            freq = calculate_freq(x, y)
            break

        # South -> West
        x = s.x - i
        y = s.y + r - i + 1
        if in_limit(x, y) and _map.is_in_sensor_range(x, y) == False:
            freq = calculate_freq(x, y)
            break

        # West -> North
        x = s.x - r + i - 1
        y = s.y - i
        if in_limit(x, y) and _map.is_in_sensor_range(x, y) == False:
            freq = calculate_freq(x, y)
            break

    if freq > 0: break

print(f"\r2. Frequency: {freq:,d}")
